### ISSUE TITLE
Add camera support

### ISSUE THREAD

**Issue by probonopd** [probonopd/WirelessPrinting#94]
This is blocked by #29.

There are now **ESP32-CAM** modules with camera and SD card. Can be had for around USD 10 from China. Ideal for our purpose.

![](https://user-images.githubusercontent.com/2480569/53688856-63063d80-3d42-11e9-9903-4f84babb86d9.png)
![](https://user-images.githubusercontent.com/2480569/53688857-63063d80-3d42-11e9-8515-e5f3666193fc.png)

@me-no-dev has added Arduino support for it :
https://github.com/espressif/esp32-camera/issues/7#issuecomment-449624965

It seems to be working well:
> Thank you so much, worked out of the box here as well. I tried it on an ESP32-CAM with a wide angle lense I got from a TTGO T-Journal order.

**Comment by chepo92** [comment]
Wow, buying it !

**Comment by probonopd** [comment]
Internal pin connections:

![](https://user-images.githubusercontent.com/2480569/54146779-a179b780-4428-11e9-95f0-90fe2776d51b.png)

Source:
https://github.com/SeeedDocument/Outsourcing/raw/master/113990580%20ESP32-CAM/113990580%20ESP32-CAM%20Product%20Specification.pdf

**Comment by probonopd** [comment]
We may not have free pins to use Serial1 on this board? GPIO16 seems to be the only pin that is on the board and not connected to cam or SD.

**Comment by chepo92** [comment]
> We need to find a way to not get the ESP32 boot messages onto the serial line since they will make it impossible to boot the printer firmware (I have tested it!).

I'll test it too 

> We may not have free pins to use Serial1 on this board? GPIO16 seems to be the only pin that is on the board and not connected to cam or SD.

I haven't tested but I think Uarts can be mapped to any pin: https://quadmeup.com/arduino-esp32-and-3-hardware-serial-ports/

**Comment by probonopd** [comment]
Yes, but you need free pins that are not connected to the camera nor the SD card. And it looks like we are short of those...

**Comment by probonopd** [comment]
In addition to making a video stream of the printer, and taking timelapse images, we might do QR code recognition to e.g., recognize settings for different filaments based on QR codes.

https://github.com/donny681/ESP32_CAMERA_QR/

**Comment by GMagician** [comment]
I read somewhere that boot log messages may be disabled on esp32 and not on esp8266, but I forget where I read it

**Comment by fred2088** [comment]
buy one, will test it out ...

**Comment by muenan** [comment]
> 
> 
> Yes, but you need free pins that are not connected to the camera nor the SD card. And it looks like we are short of those...

Maybe you can use the LED PIN (GPIO33) and with SoftSerial (RX, TX | GPIO16, GPIO33). At the LED PIN you can solder a wire to the + side for the TX. The user (v12345vtm : https://github.com/v12345vtm/timelapse-ESP32-CAM ) has developed a timelapse feature. If you could put the project and his together, it would be ideal.

**Comment by probonopd** [comment]
Let's try it out. Unfortunately I don't have this hardware.

**Comment by fred2088** [comment]
we bought it and would like to try it out ...

**Comment by probonopd** [comment]
Looking forward to hearing from you how it goes @fred2088

**Comment by probonopd** [comment]
Have you made any progress @fred2088?

**Comment by fred2088** [comment]
thinking to merge two project 
https://github.com/v12345vtm/timelapse-ESP32-CAM

**Comment by ETE-Design** [comment]
So we continue here... Can't help coding :-( My skills is not there yet... But have you looked at this project? Mabye it can help?
https://randomnerdtutorials.com/esp32-cam-video-streaming-web-server-camera-home-assistant/

**Comment by probonopd** [comment]
> Can't help coding :-( My skills is not there yet...

It's just Arduino code - so not that complicated really. Probably mostly a matter of figuring out which pins to use.

> BTW, do you have some screenshots of your GUI?

No, why? Just upload it to your ESP board and check for yourself...

**Comment by ETE-Design** [comment]
The pins is here?
https://github.com/RuiSantosdotme/ESP32-CAM-Arduino-IDE/blob/master/ESP32-CAM-Video-Streaming/ESP32-CAM-Video-Streaming.ino

Was just asking for screenshots in general, just so more people than me can see what this is and compare to ESP3D without the need og flashing...

**Comment by probonopd** [comment]
> The pins is here?

Yes, maybe. Just try it out until you have it workning.

> Was just asking for screenshots in general, just so more people than me can see what this is and compare to ESP3D without the need og flashing...

Check this video:

https://www.youtube.com/watch?v=4YOg0ReJtec

**Comment by ETE-Design** [comment]
So no Web GUI only direct integrated to Cura? Was hoping to have something like ESP 3D, not only direct control from Cura...

**Comment by probonopd** [comment]
There is also a Web interface to upload and print Gcode.

**Comment by ETE-Design** [comment]
Okey, just not show in the video then :-)

**Comment by ETE-Design** [comment]
Just now I only have the ESP8266 to Play around with... Waiting for it to be delivered from Ali...

**Comment by probonopd** [comment]
ESP8266 is fine too!

**Comment by ETE-Design** [comment]
Just dont have a SDCard connected direct to it...

**Comment by probonopd** [comment]
No need to, you can also use SPIFFS

**Comment by probonopd** [comment]
> We need to find a way to not get the ESP32 boot messages onto the serial line since they will make it impossible to boot the printer firmware (I have tested it!).

Actually, this is easy. Tested with ESP32-CAM firmware.

* Connect pin UOR to PL2303HX green = TX
* Connect pin UOT to PL2303HX white = RX
* Connect pin 5V to PL2303HX red = 5V
* Connect  pin GND to PL2303HX black = GND
* Connect pin IO15 to GND. __Boot messages gone__

**Comment by ETE-Design** [comment]
So now that you got it booting there is not that Long way to make the cam Work also :-)

**Comment by probonopd** [comment]
Someone just needs to do it.

**Comment by Colin66** [comment]
Hi, there's a project interfacing esp32-cam to 3d printer which appears relevant here :

https://github.com/eben80/ESP32-CAM-Print-Monitor

**Comment by probonopd** [comment]
Hi @Colin66 very good, so it looks like @eben80 has figured out how to access the camera while also talking to the printer via the serial line. Hence it should be a breeze to combine it with WirelessPrinting. @Colin66 would you be interested?

**Comment by Colin66** [comment]
Hi @probonopd I'm interested in the output but unfortunately I don't have the coding skills.

**Comment by eben80** [comment]
> Hi @Colin66 very good, so it looks like @eben80 has figured out how to access the camera while also talking to the printer via the serial line. Hence it should be a breeze to combine it with WirelessPrinting. @Colin66 @eben80 would you be interested to give it a try?

Hi guys,
I'm also working with very limited skills. My solution for communicating with the Serial port is extremely inefficient, probably the right way to do it would be what @luc-github uses in ESP3D.
My solution I currently have so it just about functions... my standards are very low :D

Some pointers:
I'm using SoftwareSerial for coms with the printer, the rxbuffer needs to be increased for lengthy responses like file listings.... on the hardware side I'm using voltage limiting zener diodes and a common ground connection to do the level shifting on both RX and TX.

**Comment by probonopd** [comment]
Hi @eben80, may I be curious why you are using SoftwareSerial?

**Comment by eben80** [comment]
> Hi @eben80, may I be curious why you are using SoftwareSerial?

@luc-github just asked me the same thing today :D
In my mind, I needed it because I wanted to keep UART0 for serial debugging...
I just realized that pin14 & 15 that I am using is connected to the SD Card. 
Since I haven't been using the SD Card, I haven't had any issues.

**Comment by probonopd** [comment]
Well, in this project we need the SD card of course. But I think SoftwareSerial is not a must.

**Comment by romandoss** [comment]
I run yesterday, but I had to solder wire for uart to 16 and 17 gpio directly to esp32.
And strange things, I have two esp32-cam, on one which branded AI Thinker after uploading sketch I get infinite bootloop, on other no name module worked fine.
And I also got worked sd card slot only after that modification "hasSD = SD.begin(13, SPI, fastSD ? 50000000 : 4000000);" I'm not sure correctly this or not.
And one more thing, if I reset 3d printer by button, serial gone to forever "#TIMEOUT#"

**Comment by probonopd** [comment]
When you reset the 3d printer you also need to reset the ESP at the same time. Connect the reset lines of the printer and of the ESP together.

I don't know the answer to the other questions.

**Comment by smartin015** [comment]
Not to muddy the waters too much, but there's now an ESP32S3 module with builtin camera: https://www.seeedstudio.com/XIAO-ESP32S3-Sense-p-5639.html

This board also appears to support USB host mode (crossover with #123). I splurged on a couple, should be arriving in a week or two. When I find the time to poke at them I can report back here with observations.

**Comment by probonopd** [comment]
..."supporting SD card slot for external 32GB FAT memory" sounds good. Don't forget to order that, too. We need to store the GCODE somewhere...

**Comment by smartin015** [comment]
> ..."supporting SD card slot for external 32GB FAT memory" sounds good. Don't forget to order that, too. We need to store the GCODE somewhere...

It's actually built into the module :D

**Comment by probonopd** [comment]
I was mislead by the "external".

**Comment by smartin015** [comment]
Small update - received the XIAO modules yesterday and took one for a test drive today with the [default code](https://github.com/limengdu/SeeedStudio-XIAO-ESP32S3-Sense-camera/blob/main/CameraWebServer/CameraWebServer.ino). I could connect to wifi and stream video from the camera with zero issue. Pretty slick. I can also confirm it indeed comes with an SD card slot, although I haven't yet tried using it.

That's about as far as I'm going to test for now, until I'm finished with #123 on the S2 mini.
